src/main/java/org/breedinginsight/services/TraitService.java

```diff
@@ -59,4 +83,107 @@
     }
+
+    public List<Trait> createTraits(UUID programId, List<Trait> traits, AuthenticatedUser actingUser)
+            throws DoesNotExistException, ValidatorException {
+
+        Optional<Program> optionalProgram = programService.getById(programId);
+        if (!optionalProgram.isPresent()) {
+            throw new DoesNotExistException("Program does not exist");
+        }
+        Program program = optionalProgram.get();
+
+        Optional<User> optionalUser = userService.getById(actingUser.getId());
+        if (!optionalUser.isPresent()){
+            throw new InternalServerException("Could not find user in system");
+        }
+        User user = optionalUser.get();
+
+
+        // Get our ontology
+        Optional<ProgramOntology> programOntologyOptional = programOntologyService.getByProgramId(programId);
+        if (!programOntologyOptional.isPresent()){
+            throw new InternalServerException("Ontology does not exist for program");
+        }
+        ProgramOntology programOntology = programOntologyOptional.get();
+
+        ValidationErrors validationErrors = new ValidationErrors();
+
+        // Get our program observation levels
+        List<ProgramObservationLevel> programLevels = programObservationLevelService.getByProgramId(programId);
+        for (int i = 0; i < traits.size(); i++) {
+            Trait trait = traits.get(i);
+            if (trait.getProgramObservationLevel() != null){
+                List<ProgramObservationLevel> matchingLevels = programLevels.stream()
+                        .filter(programObservationLevel -> programObservationLevel.getName().equals(trait.getProgramObservationLevel().getName()))
+                        .collect(Collectors.toList());
+                if (matchingLevels.size() == 0) {
+                    ValidationError validationError = new ValidationError("programObservationLevel.name",
+                            "Program Observation Level does not exist.", HttpStatus.NOT_FOUND);
+                    validationErrors.addError(i, validationError);
+                } else {
+                    ProgramObservationLevel dbLevel = matchingLevels.get(0);
+                    trait.getProgramObservationLevel().setId(dbLevel.getId());
+                }
+            }
+        }
+
+        // Validate the traits
+        ValidationErrors requiredFieldErrors = TraitValidator.checkRequiredTraitFields(traits);
+        ValidationErrors dataConsistencyErrors = TraitValidator.checkTraitDataConsistency(traits);
+        ValidationErrors duplicateTraits = traitValidator.checkDuplicateTraitsExisting(traits);
+        ValidationErrors duplicateTraitsInFile = TraitValidator.checkDuplicateTraitsInFile(traits);
+        validationErrors.mergeAll(requiredFieldErrors, dataConsistencyErrors, duplicateTraits, duplicateTraitsInFile);
+
+        if (validationErrors.hasErrors()){
+            throw new ValidatorException(validationErrors);
```

Comment:
If I pass in an empty json array or even empty json object in the body of the request, I get a 422 with the following output: 
```
{
  "rowErrors": [
    {
      "rowIndex": 0,
      "errors": [
        {
          "column": "method",
          "errorMessage": "Missing method",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "scale",
          "errorMessage": "Missing scale",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "traitName",
          "errorMessage": "Missing trait name",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "description",
          "errorMessage": "Missing trait description",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "programObservationLevel.name",
          "errorMessage": "Missing trait level",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        }
      ]
    }
  ]
}
```

I think it should be a 400 like you were talking about.

Reply:
Changed and pushed